src/main/java/com/bobocode/hoverla/bring/context/ConfigBasedBeanDefinition.java

```diff
@@ -0,0 +1,166 @@
+package com.bobocode.hoverla.bring.context;
+
+import com.bobocode.hoverla.bring.annotation.Bean;
+import com.bobocode.hoverla.bring.annotation.Configuration;
+import com.bobocode.hoverla.bring.annotation.Qualifier;
+import com.bobocode.hoverla.bring.exception.BeanDefinitionConstructionException;
+import com.bobocode.hoverla.bring.exception.BeanInstanceCreationException;
+import lombok.extern.slf4j.Slf4j;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Java Configuration-based bean definition. Requires instance of class marked with {@link Configuration} instance
+ * and method withing it marked with {@link Bean} that represents this bean initialization point.
+ */
+@Slf4j
+public class ConfigBasedBeanDefinition implements BeanDefinition {
+
+    private final Object configInstance;
+    private final Method beanMethod;
+    private final String name;
+    private final Class<?> type;
+    private final Map<String, Class<?>> dependencies;
+    private Object instance;
+
+    /**
+     * During execution doesn't instantiate target bean,
+     * but parses all info such as name, type, dependencies etc. and preserves this info to be used
+     *
+     * @param configInstance instance of a class marked as {@link Configuration} that contains method representing target bean
+     * @param beanMethod     method that represents target bean and is used to instantiate target bean
+     * @throws BeanDefinitionConstructionException when config class of the instance passed is not marked as {@link Configuration}
+     *                                             or bean method is not marked as {@link Bean}
+     * @throws NullPointerException                when any of arguments passed is null
+     */
+    public ConfigBasedBeanDefinition(Object configInstance, Method beanMethod) {
+        validate(configInstance, beanMethod);
+
+        log.debug("Creating {} from method '{}'", ConfigBasedBeanDefinition.class.getSimpleName(), beanMethod);
+        this.configInstance = configInstance;
+        this.beanMethod = beanMethod;
+
+        this.name = getName(beanMethod);
+        log.trace("Bean name is '{}'", name);
+
+        this.type = getType(beanMethod);
+        log.trace("'{}' bean type is '{}'", name, type);
+
+        this.dependencies = getDependencies(beanMethod);
+        log.trace("'{}' bean dependencies are {}", name, dependencies);
+    }
+
+    private static void validate(Object configInstance, Method method) {
+        log.debug("Validating arguments passed to create ConfigBased");
+        Objects.requireNonNull(configInstance, "Configuration class instance is null");
+        Objects.requireNonNull(method, "Configuration method to create bean");
```

Comment:
@DanielChanger I guess you missed "is null" in the message :)

Reply:
Good catch